Unify build process docs for all platforms (#221)
hyperledger/iroha: docs/source/guides/build.rst

``scripts/run-iroha-dev.sh`` script:

.. code-block:: shell

  bash scripts/run-iroha-dev.sh

.. hint:: Please make sure that Docker is running before executing the script.
  MacOS users could find a Docker icon in system tray, Linux users can use
  ``systemctl start docker``

After you execute this script, the following things will happen:

#. The script will check whether you have containers with Iroha already running.
Successful completion finishes with the new container shell.

#. The script will download ``hyperledger/iroha:develop-build`` and ``postgres`` images.
``hyperledger/iroha:develop-build`` image contains all development dependencies and is
based on top of ``ubuntu:16.04``. ``postgres`` image is required for starting
and running Iroha.

#. Two containers are created and launched.

#. The user is attached to the interactive environment for development and
testing with ``iroha`` folder mounted from the host machine. Iroha folder
is mounted to ``/opt/iroha`` in Docker container.

Now your are ready to build Iroha! Please go directly to `Building Iroha <#build-process>`_ section.

.. _linux-pre:

Linux
^^^^^

To build Iroha, you will need the following packages:

> ``build-essential`` ``automake`` ``libtool`` ``libssl-dev`` ``zlib1g-dev``

Reviewer: Pretty sure that this list can be reduced to `build-essential, git, cmake, ninja-build, curl, unzip, tar`

``libc6-dbg`` ``golang`` ``git`` ``tar`` ``gzip`` ``ca-certificates``
``wget`` ``curl`` ``file`` ``unzip`` ``python`` ``cmake``

Use this code to install environment dependencies on Debian-based Linux distro.

.. code-block:: shell

  apt-get update; \
  apt-get -y --no-install-recommends install \
  build-essential automake libtool \
  libssl-dev zlib1g-dev \

>   libc6-dbg golang ninja-build \

Reviewer: Same as above here.

  libraries, please consider installing the
  `latest release <https://cmake.org/download/>`_ of CMake.

Now you are ready to `install Iroha dependencies <#installing-dependencies-with-vcpkg-dependency-manager>`_.

.. _macos-pre:

MacOS
^^^^^

If you want to build Iroha from scratch and actively develop it, please use the following code
to install all environment dependencies with Homebrew:

.. code-block:: shell

  xcode-select --install

>   brew install cmake autoconf automake libtool golang ninja

Reviewer: Same as above for system packages: `cmake ninja git gcc6` should work

.. hint:: To install the Homebrew itself please run

  ``ruby -e "$(curl -fsSL https://raw.githubusercontent.com/homebrew/install/master/install)"``

Now you are ready to `install Iroha dependencies <#installing-dependencies-with-vcpkg-dependency-manager>`_.

.. _windows-pre:

Windows
^^^^^^^

.. note:: All the listed commands are designed for building 64-bit version of Iroha.

Chocolatey Package Manager
""""""""""""""""""""""""""

First of all you need Chocolatey package manager installed.
Please refer `the guide <https://chocolatey.org/install>`_ for chocolatey installation.

Building the Toolset
""""""""""""""""""""

Install CMake, Git, Microsoft compilers via chocolatey being in Administrative mode of command prompt:

.. code-block:: shell

>   choco install cmake git visualstudio2019-workload-vctools

Reviewer: ```suggestion
  choco install cmake git visualstudio2019-workload-vctools ninja
```

    # Don't forget the password you set!

Install build dependencies with Vcpkg Dependency Manager
--------------------------------------------------------

Currently we use Vcpkg as a dependency manager for all platforms - Linux, Windows and MacOS. We use a fixed version of
Vcpkg to ensure the patches we need will work. The version can be found inside the Iroha repository so we need to
clone Iroha first.
The whole process is pretty similar for all platforms but the exact commands are slightly different.

Linux and MacOS
^^^^^^^^^^^^^^^

Run in terminal:

.. code-block:: shell

  git clone https://github.com/hyperledger/iroha.git
  git clone https://github.com/Microsoft/vcpkg.git
  cd vcpkg
  git checkout $(cat ../iroha/vcpkg/VCPKG_COMMIT_SHA)
  for i in ../iroha/vcpkg/patches/*.patch; do git apply $i; done;
  ./bootstrap-vcpkg.sh
  ./vcpkg install $(cat ../iroha/vcpkg/VCPKG_DEPS_LIST | cut -d':' -f1 | tr '\n' ' ')

>   ./vcpkg install --head $(cat ../iroha/vcpkg/VCPKG_HEAD_DEPS_LIST | cut -d':' -f1 | tr '\n' ' ')

Reviewer 2: I believe this script should be placed in a separate file like prepare_vcpkg.sh and used both in docker and by dockerless Linux users. Otherwise the code is duplicated in the first and there is too much copy-paste for the second.

>   ./vcpkg integrate install

Reviewer 2: is this step needed?

Author: The last command outputs a line like `-DCMAKE_TOOLCHAIN_FILE=/Users/konstantinmunichev/src/vcpkg-my/scripts/buildsystems/vcpkg.cmake`, without it you don't know the path to .cmake file.

Reviewer 2: okay, sounds good

> Cloning the Repository

Reviewer 2: I'd move this block to environment preparation, somewhere before the `vcpkg` step, because it does not apply for
- docker users, because they must have cloned it to get the container
- all other users, because they cloned it in the `vcpkg` step.

Reviewer 3: In new structure (that I've sent yesterday) the whole part will be in "Prepare" section

Building Iroha
^^^^^^^^^^^^^^

To build Iroha, use those commands

.. code-block:: shell

>   cmake -H. -Bbuild -DCMAKE_TOOLCHAIN_FILE=/path/to/vcpkg/scripts/buildsystems/vcpkg.cmake -G "Ninja"

Reviewer 2: Is `Ninja` set up in the prerequisites preparation step?

  cmake --build build

.. note:: When building on Windows do not execute this from the Power Shell. Better use x64 Native tools command prompt.

> .. note:: You can build in multiple threads to speed the process up by a command ``cmake --build build -- -j$(nproc)``

Reviewer 2: To avoid multiple exceptions to this rule, I'd make it more general, like
```suggestion
.. note:: You can build in multiple threads to speed the process up by a command ``cmake --build build -- -j <number of threads>``
```
and then giving instructions on how to choose the number of threads depending on the number of processors and how to find the later on different platforms.

Building Iroha
^^^^^^^^^^^^^^

To build Iroha, use these commands:

.. code-block:: shell

  cmake -H. -Bbuild -DCMAKE_TOOLCHAIN_FILE=/path/to/vcpkg/scripts/buildsystems/vcpkg.cmake -G "Ninja"
  cmake --build build --target irohad -- -j<number of threads>

.. note:: On Docker the path to a toolchain file is ``/opt/dependencies/scripts/buildsystems/vcpkg.cmake``. In other
  environment please use the path you have got in previous steps.

Number of threads will be defined differently depending on the platform:
- On Linux: via ``nproc``.
- On MacOS: with ``sysctl -n hw.ncpu``.
- On Windows: use ``echo %NUMBER_OF_PROCESSORS%``.

.. note:: When building on Windows do not execute this from the Power Shell. Better use x64 Native tools command prompt.

Now Iroha is built. Although, if you like, you can build it with additional parameters described below.

CMake Parameters
^^^^^^^^^^^^^^^^

> We use CMake to build platform-dependent build files.

Reviewer: ```suggestion
We use CMake to generate platform-dependent build files.
```
